# Lab 1.1 - Cleaning and Analyzing Warehouse Data

## Objective
1. Import a CSV dataset using `pandas`
2. Clean the dataset using multiple techniques
3. Answer questions about using the data. 

*Use the challenge dataset to practice cleaning with regex.*

## Rubric

- 6 pts - Contains all required components and uses professional language
- 5 pts - Contains all required components, but uses unprofessional language, formating, etc. 
- 4 pts - Contains some, but not all, of the required components
- 3 pts - Did not submit

## Pandas reference

In addtion to the methods you learned last lab, you may need the following: 

- `df.dtypes` - the data types of the each column
- `df.columns` - the name of each column
- `df.shape` - the number of rows and columns of the dataset
- `df.dropna()` - removes null values
- `df.isnull()` - returns a dataframe with true/false values based on if the value is null. Can be used in conjuction with `.sum()`
- `pd.todatetime(...)` - sets a given column to a `datetime` object
- `df.column.isunique` - returns True if all data are unique
- `df.column.unique()` - returns all of the uniue values of a column
- `df['col'].replace(...,...)` - replaces values in a column with another value
- `df['col'].value_counts()` - returns the value counts of each unique value
- `df['col'].quantile(0.25)` or `.quartile(0.75) ` - returns the value at 25% or 75% precent    

## Preparing Data

In this section, we will import and prepare our data to be manipulated

Import `pandas` and `matplotlib.pyplot` here.

In [324]:
import pandas as pd
import matplotlib as plt

Use `pandas`' `read_csv()` to create a Dataframe called `warehouse`.

In [325]:
df = pd.read_csv("warehouse_data_challenge.csv")
display(df)

order_id   customer_id        date nearest_warehouse  \
0    ORD182494  ID6197211592  2019-06-22          Thompson   
1    ORD395518  ID0282825849  2019-12-29          Thompson   
2    ORD494479  ID0579391891  2019-03-02         Nickolson   
3    ORD019224  ID4544561904  2019-01-12         Nickolson   
4    ORD104032  ID6231506320  2019-11-28         Nickolson   
..         ...           ...         ...               ...   
495  ORD475510  ID0247417082  2019-08-07         Nickolson   
496  ORD086060  ID0130015098  03-28-2019         Nickolson   
497  ORD079320  ID2707985442  2019-10-16         Nickolson   
498  ORD026546  ID0060088412  2019-05-19         Nickolson   
499  ORD085447  ID5257193650  2019-05-25            Bakers   

                                         shopping_cart  order_price  \
0    [('Lucent 330S', 1), ('Thunder line', 2), ('iS...        12200   
1          [('Thunder line', 1), ('Universe Note', 2)]         9080   
2                 [('Thunder line', 1), ('pearTV', 2)]        10670   
3    [('Universe Note', 1), ('Alcon 10', 2), ('Oliv...        24800   
4    [('Universe Note', 1), ('Olivia x460', 1), ('i...         9145   
..                                                 ...          ...   
495             [('Alcon 10', 2), ('iAssist Line', 2)]       386776   
496  [('pearTV', 2), ('Candle Inferno', 2), ('Toshi...        22120   
497  [('pearTV', 1), ('Alcon 10', 1), ('Lucent 330S...        16490   
498  [('iStream', 1), ('Candle Inferno', 1), ('iAss...         5030   
499  [('iStream', 1), ('Toshika 750', 1), ('Alcon 1...        22370   

     delivery_charges  customer_lat  customer_long  coupon_discount  \
0               79.89    -37.815105     144.932843               10   
1               62.71    -37.802736     144.951118                0   
2               65.87    -37.821302     144.957581               10   
3               57.61    -37.811416     144.973073               15   
4               75.54     37.823859     144.969892               25   
..                ...           ...            ...              ...   
495             59.70    -37.817662     144.970405               10   
496             62.29    -37.815469     144.968083                0   
497            106.98    -37.805979     144.966101                0   
498             79.89     37.820071     144.984435               10   
499             64.62    -37.812231     144.988679               15   

     order_total  season  is_expedited_delivery  \
0       11059.89  Winter                   True   
1        9142.71  Summer                  False   
2        9668.87  Autumn                  False   
3       21137.61  Summer                  False   
4        6934.29  Spring                  False   
..           ...     ...                    ...   
495     20174.70  Winter                   True   
496     22182.29  Autumn                   True   
497     16596.98  Spring                   True   
498      4606.89  Autumn                   True   
499     19079.12  Autumn                   True   

     distance_to_nearest_warehouse  \
0                           1.2800   
1                           1.1621   
2                           1.0949   
3                           0.8571   
4                           0.5867   
..                             ...   
495                         0.1282   
496                         0.3711   
497                         1.4368   
498                         1.3191   
499                         0.6277   

                                latest_customer_review  is_happy_customer  
0    perfect phone and trusted seller. phone itself...               True  
1    it keeps dropping calls the wifi don't work th...              False  
2              five stars this is a great cheap phone.               True  
3          charger did not fit the charger didn't fit.              False  
4                                      four stars good               True  
..                           

Display the first 5 rows of `warehouse`.

In [326]:
df.head(5)

order_id   customer_id        date nearest_warehouse  \
0  ORD182494  ID6197211592  2019-06-22          Thompson   
1  ORD395518  ID0282825849  2019-12-29          Thompson   
2  ORD494479  ID0579391891  2019-03-02         Nickolson   
3  ORD019224  ID4544561904  2019-01-12         Nickolson   
4  ORD104032  ID6231506320  2019-11-28         Nickolson   

                                       shopping_cart  order_price  \
0  [('Lucent 330S', 1), ('Thunder line', 2), ('iS...        12200   
1        [('Thunder line', 1), ('Universe Note', 2)]         9080   
2               [('Thunder line', 1), ('pearTV', 2)]        10670   
3  [('Universe Note', 1), ('Alcon 10', 2), ('Oliv...        24800   
4  [('Universe Note', 1), ('Olivia x460', 1), ('i...         9145   

   delivery_charges  customer_lat  customer_long  coupon_discount  \
0             79.89    -37.815105     144.932843               10   
1             62.71    -37.802736     144.951118                0   
2             65.87    -37.821302     144.957581               10   
3             57.61    -37.811416     144.973073               15   
4             75.54     37.823859     144.969892               25   

   order_total  season  is_expedited_delivery  distance_to_nearest_warehouse  \
0     11059.89  Winter                   True                         1.2800   
1      9142.71  Summer                  False                         1.1621   
2      9668.87  Autumn                  False                         1.0949   
3     21137.61  Summer                  False                         0.8571   
4      6934.29  Spring                  False                         0.5867   

                              latest_customer_review  is_happy_customer  
0  perfect phone and trusted seller. phone itself...               True  
1  it keeps dropping calls the wifi don't work th...              False  
2            five stars this is a great cheap phone.               True  
3        charger did not fit the charger didn't fit.              False  
4                                    four stars good               True

Display the column names of the Dataframe. 

In [327]:
print(list(df.columns))

['order_id', 'customer_id', 'date', 'nearest_warehouse', 'shopping_cart', 'order_price', 'delivery_charges', 'customer_lat', 'customer_long', 'coupon_discount', 'order_total', 'season', 'is_expedited_delivery', 'distance_to_nearest_warehouse', 'latest_customer_review', 'is_happy_customer']


How many data points are in our table?

In [328]:
print(f"There are {len(df)} data points in the table")

There are 500 data points in the table


We can use one command to answer these three questions: 
- What is the average coupon discount?
- What is the largest order price?
- What is the smallest distance to warehouse?

## Cleaning Data

**Data cleaning** is the process of removing errors and inaccuracies from data to make it ready for analysis In these next steps, we will explore and clean our dataset. 


Check the dataset for null(or empty) values. How many null values are there and in which columns?

In [329]:
# displaying the totas of each column's null values
print(df.isnull().sum())

order_id                         0
customer_id                      0
date                             0
nearest_warehouse                0
shopping_cart                    0
order_price                      0
delivery_charges                 0
customer_lat                     0
customer_long                    0
coupon_discount                  0
order_total                      0
season                           8
is_expedited_delivery            0
distance_to_nearest_warehouse    0
latest_customer_review           1
is_happy_customer                0
dtype: int64


So `season` and `latest_customer_review` have missing values. Drop the rows with missing values. You should have 491 rows remaining.

In [330]:
# removing the null rows and altering the data frame without those rows
df.dropna(inplace=True)
display(df)

order_id   customer_id        date nearest_warehouse  \
0    ORD182494  ID6197211592  2019-06-22          Thompson   
1    ORD395518  ID0282825849  2019-12-29          Thompson   
2    ORD494479  ID0579391891  2019-03-02         Nickolson   
3    ORD019224  ID4544561904  2019-01-12         Nickolson   
4    ORD104032  ID6231506320  2019-11-28         Nickolson   
..         ...           ...         ...               ...   
495  ORD475510  ID0247417082  2019-08-07         Nickolson   
496  ORD086060  ID0130015098  03-28-2019         Nickolson   
497  ORD079320  ID2707985442  2019-10-16         Nickolson   
498  ORD026546  ID0060088412  2019-05-19         Nickolson   
499  ORD085447  ID5257193650  2019-05-25            Bakers   

                                         shopping_cart  order_price  \
0    [('Lucent 330S', 1), ('Thunder line', 2), ('iS...        12200   
1          [('Thunder line', 1), ('Universe Note', 2)]         9080   
2                 [('Thunder line', 1), ('pearTV', 2)]        10670   
3    [('Universe Note', 1), ('Alcon 10', 2), ('Oliv...        24800   
4    [('Universe Note', 1), ('Olivia x460', 1), ('i...         9145   
..                                                 ...          ...   
495             [('Alcon 10', 2), ('iAssist Line', 2)]       386776   
496  [('pearTV', 2), ('Candle Inferno', 2), ('Toshi...        22120   
497  [('pearTV', 1), ('Alcon 10', 1), ('Lucent 330S...        16490   
498  [('iStream', 1), ('Candle Inferno', 1), ('iAss...         5030   
499  [('iStream', 1), ('Toshika 750', 1), ('Alcon 1...        22370   

     delivery_charges  customer_lat  customer_long  coupon_discount  \
0               79.89    -37.815105     144.932843               10   
1               62.71    -37.802736     144.951118                0   
2               65.87    -37.821302     144.957581               10   
3               57.61    -37.811416     144.973073               15   
4               75.54     37.823859     144.969892               25   
..                ...           ...            ...              ...   
495             59.70    -37.817662     144.970405               10   
496             62.29    -37.815469     144.968083                0   
497            106.98    -37.805979     144.966101                0   
498             79.89     37.820071     144.984435               10   
499             64.62    -37.812231     144.988679               15   

     order_total  season  is_expedited_delivery  \
0       11059.89  Winter                   True   
1        9142.71  Summer                  False   
2        9668.87  Autumn                  False   
3       21137.61  Summer                  False   
4        6934.29  Spring                  False   
..           ...     ...                    ...   
495     20174.70  Winter                   True   
496     22182.29  Autumn                   True   
497     16596.98  Spring                   True   
498      4606.89  Autumn                   True   
499     19079.12  Autumn                   True   

     distance_to_nearest_warehouse  \
0                           1.2800   
1                           1.1621   
2                           1.0949   
3                           0.8571   
4                           0.5867   
..                             ...   
495                         0.1282   
496                         0.3711   
497                         1.4368   
498                         1.3191   
499                         0.6277   

                                latest_customer_review  is_happy_customer  
0    perfect phone and trusted seller. phone itself...               True  
1    it keeps dropping calls the wifi don't work th...              False  
2              five stars this is a great cheap phone.               True  
3          charger did not fit the charger didn't fit.              False  
4                                      four stars good               True  
..                           

Now let's take a look at the data types of each column and see if they make sense.

In [331]:
# checking column types
df.dtypes

order_id                          object
customer_id                       object
date                              object
nearest_warehouse                 object
shopping_cart                     object
order_price                        int64
delivery_charges                 float64
customer_lat                     float64
customer_long                    float64
coupon_discount                    int64
order_total                      float64
season                            object
is_expedited_delivery               bool
distance_to_nearest_warehouse    float64
latest_customer_review            object
is_happy_customer                   bool
dtype: object

Data type of date column is object instead of datetime. So fix this problem and check the datatypes again.

In [332]:
# overwriting the date column with the formated dates
df["date"] = pd.to_datetime(df["date"], format="mixed")

# checking column types
df.dtypes

order_id                                 object
customer_id                              object
date                             datetime64[ns]
nearest_warehouse                        object
shopping_cart                            object
order_price                               int64
delivery_charges                        float64
customer_lat                            float64
customer_long                           float64
coupon_discount                           int64
order_total                             float64
season                                   object
is_expedited_delivery                      bool
distance_to_nearest_warehouse           float64
latest_customer_review                   object
is_happy_customer                          bool
dtype: object

Now let's check for errors. 

First, let's make sure that our data in `order_id` is unique and we don't have any duplicates. 

In [333]:
df["order_id"].is_unique

True

Now, let's check what the unique values of `nearest_warehouse` are.

In [334]:
print(list(df["nearest_warehouse"].unique()))

['Thompson', 'Nickolson', 'Bakers', 'nickolson', 'thompson']


Thompson and thompson should be the same, but here they are being counted as different warehouse. Similarly Nickolson and nickolson are same. Change all of the lowercase names to Title case. 

In [335]:
df["nearest_warehouse"] = df["nearest_warehouse"].str.title()
warehouses = list(df["nearest_warehouse"].unique())
print(warehouses)

['Thompson', 'Nickolson', 'Bakers']


Check the amount of times each `nearest_warehouse` is said.

In [336]:
for warehouse in warehouses: 
    print(f"The amount of \"{warehouse}\" warehouses is: {df["nearest_warehouse"].value_counts().get(warehouse, 0)}")

The amount of "Thompson" warehouses is: 192
The amount of "Nickolson" warehouses is: 183
The amount of "Bakers" warehouses is: 116


The value counts should be Thompson: 192, Nickolson: 183, Bakers: 116

Check the counts and unique values of `is_happy_customer`.

In [337]:
happyness = list(df["is_happy_customer"].unique())
for emotion in happyness:
    print(f"The amount of {"dissatisfied" if emotion == True else "satisfied"} customers is {df["is_happy_customer"].value_counts().get(emotion, 0)}")

The amount of dissatisfied customers is 352
The amount of satisfied customers is 139


Create a table with `order_price`, `delivery_charges`, `coupon_discount`, and `order_total`. Look at the first 10 rows. Is the `order_total` column correct?

In [338]:
pd.options.mode.copy_on_write = True
df10 = df.head(10)
df10["true_total"] = df['order_price'] + df["delivery_charges"] - df["coupon_discount"]
print(df10["true_total"] == df10["order_total"])
display(df10)

0     False
1      True
2     False
3     False
4     False
5     False
6     False
9     False
10     True
11    False
dtype: bool


order_id   customer_id       date nearest_warehouse  \
0   ORD182494  ID6197211592 2019-06-22          Thompson   
1   ORD395518  ID0282825849 2019-12-29          Thompson   
2   ORD494479  ID0579391891 2019-03-02         Nickolson   
3   ORD019224  ID4544561904 2019-01-12         Nickolson   
4   ORD104032  ID6231506320 2019-11-28         Nickolson   
5   ORD146760  ID0311654900 2019-09-16            Bakers   
6   ORD337984  ID3394768956 2019-09-14          Thompson   
9   ORD462194  ID5301568579 2019-03-21          Thompson   
10  ORD034800  ID4283908179 2019-08-03            Bakers   
11  ORD361636  ID0589500304 2019-12-05         Nickolson   

                                        shopping_cart  order_price  \
0   [('Lucent 330S', 1), ('Thunder line', 2), ('iS...        12200   
1         [('Thunder line', 1), ('Universe Note', 2)]         9080   
2                [('Thunder line', 1), ('pearTV', 2)]        10670   
3   [('Universe Note', 1), ('Alcon 10', 2), ('Oliv...        24800   
4   [('Universe Note', 1), ('Olivia x460', 1), ('i...         9145   
5         [('Thunder line', 2), ('Universe Note', 1)]         7810   
6   [('Candle Inferno', 1), ('Alcon 10', 1), ('Tos...        13700   
9   [('Universe Note', 1), ('Lucent 330S', 1), ('T...        13320   
10  [('Alcon 10', 2), ('pearTV', 2), ('iStream', 1...        31895   
11                [('Lucent 330S', 1), ('pearTV', 2)]        13850   

    delivery_charges  customer_lat  customer_long  coupon_discount  \
0              79.89    -37.815105     144.932843               10   
1              62.71    -37.802736     144.951118                0   
2              65.87    -37.821302     144.957581               10   
3              57.61    -37.811416     144.973073               15   
4              75.54     37.823859     144.969892               25   
5              71.22     37.820245     145.014944               10   
6              74.84    -37.807743     144.951570                5   
9              62.26    -37.808675     144.942337               15   
10             78.25    -37.811334     145.008678                0   
11             77.29    -37.820228     144.957411               25   

    order_total  season  is_expedited_delivery  distance_to_nearest_warehouse  \
0      11059.89  Winter                   True                         1.2800   
1       9142.71  Summer                  False                         1.1621   
2       9668.87  Autumn                  False                         1.0949   
3      21137.61  Summer                  False                         0.8571   
4       6934.29  Spring                  False                         0.5867   
5       7100.22  Spring                  False                         2.0752   
6      13089.84  Spring                  False                         0.6767   
9      11384.26  winter                   True                         0.6093   
10     31973.25  Winter                   True                         1.1919   
11     10464.79  Summer                  False                         1.0829   

                               latest_customer_review  is_happy_customer  \
0   perfect phone and trusted seller. phone itself...               True   
1   it keeps dropping calls the wifi don't work th...              False   
2             five stars this is a great cheap phone.               True   
3         charger did not fit the charger didn't fit.              False   
4                                     four stars good               True   
5   stolen phone sold us a stolen phone so we coul...              False   
6   love our inferno stick,easy to set up and have...               True   
9   does not live up to its reputation. customer s...              False   
10  i love this phone it is so user friendly and t...               True   
11  great phone great deal! phone in fantastic con...               True   

    true_total  
0     12269.89  
1      9142.71  
2     10725.87  
3    

### Remove Outliers

We are going to remove statistical outliers in `order_total`. To remove outliers, follow this process:

- Find upper and lower quartiles
- Find IQR range
- Find upper and lower limit
- Remove elements 

In [339]:
quarters = df.order_total.quantile([0.25, 0.75])

In [340]:
quarters = list(df.order_total.quantile([0.25, 0.75]))
upper_quarter = quarters[1]
lower_quarter = quarters[0]
IQR = upper_quarter - lower_quarter
print(f"{round(IQR, 2)}")

11795.82


In [341]:
min = df.order_total.min()
max = df.order_total.max()
print(f"The lower limit is {min}", f"The Upper limit is {max}", sep="\n")

The lower limit is 639.29
The Upper limit is 5688269.6


In [ ]:
dropped_outliers = df[(df["order_total"] < lower_quarter - 1.5 * IQR) | (df["order_total"] > upper_quarter + 1.5 * IQR)].index
df.drop(dropped_outliers, inplace= True)
display(df)

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

There are 18 records with outliers to remove. 

## Explore Data

Now that our data is clean, we can answer questions about it!

What is the largest order?

What is the smallest order?

Without outliers, what are the top 5 most expensive orders?

What warehouse is the closest to most customers? Create a visualization.

What percentage of customers are satisfied with their purchase? Create a visualization.